src/hotspot/share/nmt/memMapPrinter.cpp

```diff
@@ -0,0 +1,291 @@
+// those that have been used with mmap. Flags left out are printed with their nmt flag name.
+#define NMT_FLAGS_DO(f) \
+  /* flag, short, description */ \
+  f(mtGCCardSet,      "CARDTBL", "GC Card table") \
+  f(mtClassShared,    "CDS", "CDS archives") \
+  f(mtClass,          "CLASS", "Class Space") \
+  f(mtCode,           "CODE", "Code Heap") \
+  f(mtGC,             "GC", "GC support data (e.g. bitmaps)") \
+  f(mtInternal,       "INTERN", "Internal") \
+  f(mtJavaHeap,       "JAVAHEAP", "Java Heap") \
+  f(mtOther,          "JDK", "allocated by JDK libraries other than VM") \
+  f(mtMetaspace,      "META", "Metaspace nodes (non-class)") \
+  f(mtSafepoint,      "POLL", "Polling pages") \
+  f(mtThreadStack,    "STACK", "(known) Thread Stack") \
+  f(mtTest,           "TEST", "JVM internal test mappings")
+  //end
+
+static const char* get_shortname_for_nmt_flag(MEMFLAGS f) {
+#define DO(flag, shortname, text) if (flag == f) return shortname;
+  NMT_FLAGS_DO(DO)
+#undef DO
+  return NMTUtil::flag_to_enum_name(f);
+}
+
+/// NMT virtual memory
+
+static bool range_intersects(const void* from1, const void* to1, const void* from2, const void* to2) {
+  const void* const min = MAX2(from1, from2);
+  const void* const max = MIN2(to1, to2);
+  return min < max;
+}
+
+// A Cache that correlates range with MEMFLAG, optimized to be iterated quickly
+// (cache friendly).
+class CachedNMTInformation : public VirtualMemoryWalker {
+  struct Range { const void* from; const void* to; };
+  // Unfortunately, we need to allocate manually, raw, since we must prevent
+  // NMT deadlocks (ThreadCritical).
+  Range* _ranges;
+  MEMFLAGS* _flags;
+  uintx _count, _capacity;
+public:
+  CachedNMTInformation() : _ranges(nullptr), _flags(nullptr), _count(0), _capacity(0) {}
+
+  ~CachedNMTInformation() {
+    ALLOW_C_FUNCTION(free, ::free(_ranges);)
+    ALLOW_C_FUNCTION(free, ::free(_flags);)
+  }
+
+  bool add(const void* from, const void* to, MEMFLAGS f) {
+    assert(_count == 0 || (from >= _ranges[_count - 1].to), "NMT regions unordered?");
+    // we can just fold two regions if they are adjacent and have the same flag.
+    if (_count > 0 && from == _ranges[_count - 1].to && f == _flags[_count - 1]) {
+      _ranges[_count - 1].to = to;
+      return true;
+    }
+    if (_count == _capacity) {
+      // Enlarge if needed
+      const uintx new_capacity = MAX2((uintx)4096, 2 * _capacity);
+      ALLOW_C_FUNCTION(realloc, _ranges = (Range*)::realloc(_ranges, new_capacity * sizeof(Range));)
+      ALLOW_C_FUNCTION(realloc, _flags = (MEMFLAGS*)::realloc(_flags, new_capacity * sizeof(MEMFLAGS));)
+      if (_ranges == nullptr || _flags == nullptr) {
+        // In case of OOM lets make no fuzz. Just return.
+        return false;
+      }
+      _capacity = new_capacity;
+    }
+    assert(_capacity > _count, "Sanity");
+    _ranges[_count].from = from;
+    _ranges[_count].to = to;
```

Comment:
Or just `_ranges[_count] = Range{from, to}`

Reply:
ok